### ISSUE TITLE
Take screenshot with full page size

### ISSUE THREAD

**Issue by Grawl** [DevExpress/testcafe#1520]
### Are you requesting a feature or reporting a bug?

Requesting a feature

Tested page URL: any

Test code

```js
// as you wish
```

### Specify your

* operating system: macOS
* testcafe version: 0.15.0
* node.js version: 8.0.0

**Comment by AlexanderMoskovkin** [comment]
Hi,
 
Do you mean taking a screenshot of the whole page, not only of its visible part (e.g., if the page has a large scrollbar height, you'd like to see the whole page in a screenshot)? Do I understand you right?

**Comment by Grawl** [comment]
Yes

**Comment by adamorlowskipoland** [comment]
I'm looking for the same usibility.
I have found that for now taking fullpage screenshots is available only on firefox.

https://saucelabs.com/blog/Selenium-Tips-Capturing-Screenshots-vs-Scrollbars

**Comment by HenningThiemann** [comment]
@AlexanderMoskovkin did you implement this feature already?

**Comment by HenningThiemann** [comment]
@AlexanderMoskovkin thanks for your answer. Would be a great feature IMO

**Comment by rquellh** [comment]
@AlexanderMoskovkin I'm also interested in this feature. Thanks for looking into it.

**Comment by zhaoace** [comment]
@AlexanderMoskovkin  We're looking forward to this feature very much. Thanks for planing it.

**Comment by AndreyBelym** [comment]
Hi @marcin-chwedczuk-meow, thanks for feedback! You can automatically take screenshots on failures using [the corresponding option](https://devexpress.github.io/testcafe/documentation/using-testcafe/command-line-interface.html#-s---screenshots-on-fails), but it will take only the visible part of a page? of course. Could you please explain, what useful or important things can be in the _invisible_ part of a page? Could you share your experience?

**Comment by marcin-chwedczuk-meow** [comment]
One use-case that comes to my mind is that captured area of the screen does not contain all useful information. Like e.g. let's say we have a page with all invoice data and for some reasons a particular test fails at the bottom of the page. I would appreciate if I see not only part of the page where test failed but also all other invoice data that may be on the screen (my test is concentrating on a business side of the things - so I am not interested that e.g. button is disabled but why it is disabled). So indeed for acceptance tests this may be useful.

**Comment by jantonypdx** [comment]
Another use case is Visual Validation testing:
You write a Testcafe-driven workflow that visits many pages. On each page, it capture a screenshot of the full page and then sends it to a visual validation tool. This tool looks at the image & compares it to a previous "baseline" image. It then returns a "difference" score (e.g. 0.00 == images identical; 0.01+ - 100.0 == % change). You then write assertions that fail the test above a threshold value.

Examples of visual validation tools include [Resemble.js](https://www.npmjs.com/package/resemblejs), [Applitools](https://www.youtube.com/watch?v=-jiQdVzuenM), or [Galen Framework](https://www.youtube.com/watch?v=bheFQfEGR6U) (no affiliation with these). This makes it easier for people to quickly know & see which pages have changed on their websites (especially when you support _many_ different sites & locales).

There are other automation frameworks that let you capture full-screen screenshots, but I'd rather use TestCafe than have to support two different frameworks only for including this type of testing. 

Thanks for considering this feature!

**Comment by moffsugita** [comment]
I hope. `.resizeWindow` can't

**Comment by alexschwantes** [comment]
This is a really important feature for us as often the item to be verified is further down on the page which doesn't show on the screenshot. For example if we have a list of a lot of items on a page and try to validate if a particular one exists, if the test fails we just get the top part of the screen and so can't visually verify that it didn't exist further down the page. While the whole page isn't in the visible viewport, I wouldn't class content further down the page as being 'invisible' as this suggests that it is completely 'hidden' from view.

Also coming from selenium it feels like an expected feature to have.

**Comment by fraalvper** [comment]
I add `--window-size='1360,768'` in my package.json and it worked for me.

full line
`"test-chrome-report": "./node_modules/.bin/cucumber-js  --format json:./reports/report.json --world-parameters \"{\\\"browser\\\": \\\"chrome --window-size='1360,768'\\\"}\""`

**Comment by bdjos1** [comment]
I'm also experiencing this issue where the entire screenshot of a page is not being taken in desktop, only a portion of it.

@fraalvper It would be great if you could provide more details on this workaround. In the command you have shown, it is referring to tests run on cucumber. Are you using gherkin-testcafe or some tool like that instead of testcafe?

**Comment by Pr60** [comment]
Can anyone suggest a workaround for this until this is done? I tried t.resizeWindow() but this isn't working for me on `chrome:headless`. I am looking for a workaround for headless mode.

**Comment by AndreyBelym** [comment]
@Pr60, have you tried to use the [emulation mode](https://devexpress.github.io/testcafe/documentation/using-testcafe/common-concepts/browsers/using-chrome-device-emulation.html)? You can emulate a device with the screen size that is equal to the maximum size of your page.

**Comment by fralvarop** [comment]
@bdjos1 Yes im using a suit with testcafe + cucumber. This is my package.json

https://github.com/fralvarop/testcafe-cucumberjs/blob/master/package.json

**Comment by Pr60** [comment]
@AndreyBelym @fralvarop - Screenshot problem got resolved after I implemented both of yours solution but some of my tests started failing because there are multiple scrollbars in my page which did not respond properly because of small page size on my Mac. Could there be another possible way? or do we know a timeline for the implementation of this feature?

**Comment by alexschwantes** [comment]
@AndreyBelym what is the maximum screen size of our page? It could be 1 page or 10 pages long and would possibly change during the test so I don't know if setting a hardcoded value is really a solution. If I set it to some large value, will that mean I always get a really long screenshot even if the page isn't that large?

**Comment by AlexKamaev** [comment]
@alexschwantes 
You can detect dimensions of your page using `document.clientWidht` and `document.clientHeight` in [ClientFunction](https://devexpress.github.io/testcafe/documentation/test-api/obtaining-data-from-the-client/). After that you can resize your browser by the [resize ](https://devexpress.github.io/testcafe/documentation/test-api/actions/resize-window.html)method and then call the`takeScreenshot` method. Please note that is better to do it in Emulation and Headless mode.

@Pr60 @alexschwantes 
We know how this feature is important for users, so we added it in our [Roadmap for 2019](https://devexpress.github.io/testcafe/roadmap/#api)
Since this feature is highly requested we consider it as one of our prior tasks

**Comment by byrne-greg** [comment]
@AlexKamaev 
this is an issue we're looking at right now and trying to devise a workaround. We have a mix of large and small pages, most outside the visible area. I've had no success with `document.clientWidth` or height, they seem undefined when used with the ClientFunction. Did you mean something like `ClientFunction(() => window.document.body.clientHeight)`? The problem with clientWidth or clientHeight is that it's still the size of the visible window, even in Headless mode, and not the full page.

I've found a better solution that works is to use the `scrollHeight` on the `body` element. This produces full page screenshots.
```
const pageHeight = ClientFunction(() => window.document.body.scrollHeight)
t.resizeWindow(1200, await pageHeight() );
```
However, one problem with this method is that it works only when the pages are ordered from large to small. The canvas seems 'stretched' when ordered the other (see #3648 )

**Comment by AlexKamaev** [comment]
@byrne-greg 
Thanks, for this page I have `document.body.clientHeight` equal to `document.body.scrollHeight`, but it does not matter, in any case, you got it right. I am happy to hear that this workaround works for you.
 
> However, one problem with this method is that it works only when the pages are ordered from small to large during test execution. The page seems 'stretched' when ordered the other way around

This behavior is very strange, we'll discuss it in the ticket you provided

**Comment by jiteshsojitra** [comment]
+1 for taking entire window/browser screenshot.
